python/src/opendp/extras/aim_new.py

```diff
@@ -0,0 +1,205 @@
+            q_r_array = [] # this is an array of length |C_t| where each element has n entries
+
+            t += 1 # Algo 2 line 11; goes to the next index in epsilons and sigmas array that was just added
+            rho_used += (epsilons[t]**2 / 8 + 1 / (2 * sigmas[t]**2)) # Algo 2 line 12
+            C_t, C_t_indices = zip(*[(query, index) for index, query in enumerate(W_plus)
+                                     if query not in selected_queries 
+                                     and get_new_query_fit(selected_queries, query, rho_used)]) # Algo 2 line 13
+            
+            # calculate the q_r(D) for each r in C_t; Algo 2 line 14 getting each part of the score function
+            for query, index in zip(C_t, C_t_indices):
+                w_r = marginal_weights[index]
+                M_r_D = all_marginals[index]
+                marginal_fn = make_distribution_marginal_transformation(query)
+                M_r_p_hat = marginal_fn(p_hat)
+                q_r = w_r * (np.sum(np.abs(M_r_D - M_r_p_hat)) - math.sqrt(2 / math.pi) * sigmas[t] * n_r_array[index])
+                q_r_array.append(q_r)
+
+            # SELECT the next query using exponential mechanism 
+            gumbel_scale = 2 * max(marginal_weights) / epsilons[t] # FIX_THIS
+            this_expo_mechanism = dp.m.make_report_noisy_max_gumbel(
+                input_domain = dp.vector_domain(dp.atom_domain(float)),
+                input_metric = dp.linf_distance(float),
+                scale = gumbel_scale,
+                optimize = "max"
+            )
+            selected_query_index = this_expo_mechanism(q_r_array) # use this expo mech to select the index of next query
+            selected_query_indices.append(selected_query_index)
+            selected_query = C_t[selected_query_index] # Algo 2 line 14; get the actual query from C_t
+            selected_queries.append(selected_query) # Algo 2 line 14; add newly selected query to the list
+
+            # MEASURE the marginal on selected query; Algo 2 line 15
+            gaussian = dp.m.make_gaussian(input_domain = input_domain, input_metric = input_metric, scale = sigmas[t])
+            selected_noisy_marginal = np.array([gaussian(val) for val in all_marginals[selected_query_index]]) # changed to measurement
+            selected_noisy_marginals.append(selected_noisy_marginal)
+
+            # ESTIMATE new synthetic dataset using Private-PGM; Algo 2 line 16
+            p_hat_new = np.zeros((n, d)) # FIX_THIS; get the actual new thing
+
+            # ANNEAL; Algo 2 line 17; add the new epsilon and sigma to the epsilons and sigmas lists
+            selected_query_marginal_fn = make_distribution_marginal_transformation(selected_query)
+            new_selected_marginal = selected_query_marginal_fn(p_hat_new)
+            old_selected_marginal = selected_query_marginal_fn(p_hat)
+            marginals_l1_difference = np.sum(np.abs(new_selected_marginal - old_selected_marginal))
```

Comment:
The choices of parameters are fully adaptive (dependent on prior releases), so this needs a privacy filter, not adaptive composition